main/src/main/scala/sbt/Load.scala

```diff
@@ -648,54 +676,66 @@ object Load {
-        case p: Plugins          => pluginSettings(p)
-        case p: AutoPlugins      => autoPluginSettings(p)
-        case q: Sequence         => (Seq.empty[Setting[_]] /: q.sequence) { (b, add) => b ++ expandSettings(add) }
+    log: Logger): Project =
+    timed(s"Load.resolveProject(${rawProject.id})", log) {
+      import AddSettings._
+      // 1. Apply all the project manipulations from .sbt files in order
+      val transformedProject =
+        timed(s"Load.resolveProject(${rawProject.id}): transformedProject", log) {
+          configFiles.flatMap(_.manipulations).foldLeft(rawProject) { (prev, t) =>
+            t(prev)
+          }
+        }
+      // 2. Discover all the autoplugins and contributed configurations.
+      val autoPlugins =
+        timed(s"Load.resolveProject(${rawProject.id}): autoPlugins", log) {
+          try loadedPlugins.detected.deducePluginsFromProject(transformedProject, log)
+          catch { case e: AutoPluginException => throw translateAutoPluginException(e, transformedProject) }
+        }
+      val autoConfigs = autoPlugins.flatMap(_.projectConfigurations)
+
+      // 3. Use AddSettings instance to order all Setting[_]s appropriately
+      val allSettings = {
+        // TODO - This mechanism of applying settings could be off... It's in two places now...
+        lazy val defaultSbtFiles = configurationSources(transformedProject.base)
+        // Grabs the plugin settings for old-style sbt plugins.
+        def pluginSettings(f: Plugins) =
+          timed(s"Load.resolveProject(${rawProject.id}): expandSettings(...): pluginSettings($f)", log) {
+            val included = loadedPlugins.detected.plugins.values.filter(f.include) // don't apply the filter to AutoPlugins, only Plugins
+            included.flatMap(p => p.settings.filter(isProjectThis) ++ p.projectSettings)
+          }
+        // Filter the AutoPlugin settings we included based on which ones are
+        // intended in the AddSettings.AutoPlugins filter.
+        def autoPluginSettings(f: AutoPlugins) =
+          timed(s"Load.resolveProject(${rawProject.id}): expandSettings(...): autoPluginSettings($f)", log) {
+            autoPlugins.filter(f.include).flatMap(_.projectSettings)
+          }
+        // Grab all the settigns we already loaded from sbt files
+        def settings(files: Seq[File]): Seq[Setting[_]] =
+          timed(s"Load.resolveProject(${rawProject.id}): expandSettings(...): settings($files)", log) {
+            for {
+              file <- files
+              config <- (memoSettings get file).toSeq
+              setting <- config.settings
+            } yield setting
+          }
+        // Expand the AddSettings instance into a real Seq[Setting[_]] we'll use on the project
+        def expandSettings(auto: AddSettings): Seq[Setting[_]] = auto match {
+          case BuildScalaFiles     => rawProject.settings
+          case User                => globalUserSettings.cachedProjectLoaded(loadedPlugins.loader)
```

Comment:
A few thoughts:

1. java.net.URL's hashCode is (insanely) a blocking, synchronized dns-resolving operation, making it (IMHO) a bad data-type to use in general, but especially if used as the (or a part of) the key of any hash-based collection (map/set/etc).
2. I don't think it's universally correct to define a cache of a java.net.URLClassLoader-using-function in terms of the hashset of its urls as I think the ordering is important.
3. I also don't think it's universally correct to define the cache solely in terms of its URL all together, as I think there are other properties of java.net.URLClassLoader that affect it's behaviour.
4. Finally, this is trading memory for speed, and I wonder how it impacts multi-project builds that cross-build for Scala.js, where there are already memory struggles. (eg. shapeless, circe, cats)

I say "universally" because this is just from a tunnel vision review of the change, not in terms of sbt's use here or the real-world situation behind this change.

Also note that URLClassLoader doesn't define hashCode at all, falling back on java.lang.Object generic (identity-based?) hashCode.